[docs] Explain the builtin commands and the separating script for script run stage

ffjlabo opened the pull request:
**What this PR does / why we need it**:

as title

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

- **How are users affected by this change**:
- **Is this breaking change**:
- **How to migrate (if breaking change)**:

khanhtc1202 commented on docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,51 @@ The commands run in the directory where this application configuration file exis
+- ssh
+- jq
+- curl
+- and the builtin commands installed in the base image.
```
Users can also use piped installed tools such as terraform, kubectl, helm, kustomize, etc 👀

ffjlabo replied:
@khanhtc1202 
Thank you for the comment.
I think it might be unexpected usage for these tools.
WDYT?

khanhtc1202 replied:
Sorry, I don't really get your point. Could you explain it :pray:

ffjlabo replied:
📝 Discussed about it in person. I agree that users can use the commands installed for the piped.
- we can use them on the script run stage
- but this may change in the future

ffjlabo replied:
fixed on [48d7819](https://github.com/pipe-cd/pipecd/pull/5127/commits/48d781964683100dcd86cc0f1dfecde6c2a572a7)

Warashi replied:
The tool binaries in `$PIPED_TOOLS_DIR` are installed just before it's needed.
So, the binaries are sometimes absent, such as when the SCRIPT_RUN stage is the first stage of the pipeline and the binaries are not installed yet.

khanhtc1202 replied:
I understand your both point now, and agree with @Warashi If it isn't always available, the it should not be listed. But, it could be better to mention that some tools could be provided under PIPED_TOOLS_DIR so that users can check it out by themself, wdyt? @ffjlabo @Warashi

Warashi replied:
I agree with @khanhtc1202 .
It's okay to mention that some tools may be available under PIPED_TOOLS_DIR.
But I think we also have to mention that the tools can be absent sometimes.

ffjlabo replied:
@khanhtc1202 @Warashi 
> So, the binaries are sometimes absent, such as when the SCRIPT_RUN stage is the first stage of the pipeline and the binaries are not installed yet.

I forgot it. So I think we should not explain it.

It might be confusing for users who use SCRIPT_RUN that the command does not exist until the first deployment is executed.
If the piped administrator and the app administrator are different, I don't think there is a way for the app administrator to check the dir.

WDYT?

Warashi replied:
The app administrator can use the SCRIPT_RUN stage, such as executing `ls`, to check whether the commands exist.
But I think it's nonsense to check whether the commands exist. If the administrator has to use a command, they should install it in the SCRIPT_RUN stage or ask the piped admin to use a custom container image containing the command.

I think it's okay to mention some tools are sometimes available under PIPED_TOOLS_DIR because it's a fact.
On the other hand, users must use those tools carefully, and we must explain why users must pay attention.

ffjlabo replied:
I tried adding it anyway. [bfb1765](https://github.com/pipe-cd/pipecd/pull/5127/commits/bfb176524c3e29e96c5c11ef6c870479ede2646e)

ffjlabo replied:
Plz check it 🙏 @Warashi @khanhtc1202

t-kikuc commented on docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,51 @@ The commands run in the directory where this application configuration file exis
+kind: KubernetesApp
+spec:
+  name: script-run
+  labels:
+    env: example
+    team: product
```
[nits] I think `labels` section is not necessary for the explanation.

t-kikuc commented on docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,51 @@ The commands run in the directory where this application configuration file exis
+
+The public piped image available in PipeCD main repo (ref: [Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/cmd/piped/Dockerfile)) is based on [alpine](https://hub.docker.com/_/alpine/) and only has a few UNIX command available (ref: [piped-base Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/tool/piped-base/Dockerfile)). 
+
+If you want to use your commands, you can:
```
[nits] I wanna clarify whether the options are isolated or combined.

e.g. 
```suggestion
If you want to use your commands, follow these steps:
```

t-kikuc commented on docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md:
````diff
@@ -55,6 +55,51 @@ The commands run in the directory where this application configuration file exis
+└── script.sh
+```
+
+## Builtin command
````
[nits] 
```suggestion
## Builtin commands
```

t-kikuc commented on docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,51 @@ The commands run in the directory where this application configuration file exis
+- curl
+- and the builtin commands installed in the base image.
+
+The public piped image available in PipeCD main repo (ref: [Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/cmd/piped/Dockerfile)) is based on [alpine](https://hub.docker.com/_/alpine/) and only has a few UNIX command available (ref: [piped-base Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/tool/piped-base/Dockerfile)). 
```
[nits]

```suggestion
The public piped image available in PipeCD main repo (ref: [Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/cmd/piped/Dockerfile)) is based on [alpine](https://hub.docker.com/_/alpine/) and only has a few UNIX commands available (ref: [piped-base Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/tool/piped-base/Dockerfile)). 
```

t-kikuc reviewed (commented):
Great improvement 👍 
I left some nits

ffjlabo commented:
@t-kikuc Thank you for the comment 🙏  I fixed them.

- fix description for the usecase [3ab416e](https://github.com/pipe-cd/pipecd/pull/5127/commits/3ab416e3d56ed7190f295d5e7423c06ab1f1d907)
  - clarified the options as isolated ones.
  - removed the one that is not recommended 
- remove labels [ed4f0b0](https://github.com/pipe-cd/pipecd/pull/5127/commits/ed4f0b0f2a7191edff706630235cc495d188d07b)
- wrong spell [c502a27](https://github.com/pipe-cd/pipecd/pull/5127/commits/c502a27693e5c4cad1748e6f088fbd316e13f835)

Warashi commented on docs/content/en/docs-v0.46.x/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,47 @@ The commands run in the directory where this application configuration file exis
+For example, If you are using the container platform and the offcial piped container image, you can use the command below.
+- git
+- ssh
+- jq
```
We don't install jq in 0.46.x images, because it's installed at 0.48.0 release
https://github.com/pipe-cd/pipecd/pull/5009

ffjlabo replied:
fixed on [dc78b67](https://github.com/pipe-cd/pipecd/pull/5127/commits/dc78b676c422bce4c9189394734549ed1b878193)

Warashi commented on docs/content/en/docs-v0.47.x/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,47 @@ The commands run in the directory where this application configuration file exis
+For example, If you are using the container platform and the offcial piped container image, you can use the command below.
+- git
+- ssh
+- jq
```
We don't install jq in 0.47.x images, because it's installed at 0.48.0 release
https://github.com/pipe-cd/pipecd/pull/5009

ffjlabo replied:
fixed on [dc78b67](https://github.com/pipe-cd/pipecd/pull/5127/commits/dc78b676c422bce4c9189394734549ed1b878193)

khanhtc1202 commented:
@ffjlabo Could you check this 👀

Warashi reviewed (approved):
LGTM

t-kikuc commented:
@khanhtc1202 Would you check again?

khanhtc1202 commented on docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md:
```diff
@@ -55,6 +55,52 @@ The commands run in the directory where this application configuration file exis
+Currently, you can use the commands which are installed in the environment for the piped.
+
+For example, If you are using the container platform and the offcial piped container image, you can use the command below.
+- git
+- ssh
+- jq
+- curl
+- kubectl, helm, kustomize, terraform stored in $PIPED_TOOL_DIR 
+  - Be careful. These tools are sometimes absent because they are installed when the first deployment has been done using each tool. Please check $PIPED_TOOL_DIR before you use them.
+  - Binaries can be specified as "kubectl-${version}" or "kubectl". 
+  - If no version is specified on the piped config or app.pipecd.yaml, both the ones with and without default version suffix will be installed. 
+  - Please check the [code](https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/toolregistry/install.go#L28C1-L33C2) about the default versions.
+- and the builtin commands installed in the base image.
```
```suggestion
- git
- ssh
- jq
- curl
- commands installed by piped in $PIPED_TOOL_DIR (check at runtime)
- built-in commands installed in the base image.
```

How about this? @ffjlabo

ffjlabo replied:
@khanhtc1202 
I think we also have to mention that the tools can be absent sometimes.
WDYT?

khanhtc1202 replied:
That is why I added the note `(check at runtime)`. I think it would be better not to confuse users by showing too much explanation 🤔

ffjlabo replied:
OK, I agree with your point.
I will fix it👌

ffjlabo replied:
fixed [48022da](https://github.com/pipe-cd/pipecd/pull/5127/commits/48022daafc84b0beb3e27cb973d4f9e6979a742d)

khanhtc1202 reviewed (approved):
Thank you 💯
